### PR TITLE
Stn/unfork glutin

### PR DESCRIPTION
**Pull request by stuartnelson3**
continuation of work started in https://github.com/servo/servo/pull/17311

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because it should be a pure refactoring.

Build is currently working on Linux and mac, need to check other platforms.

Note:

PR is currently blocked by https://github.com/tomaka/winit/pull/349

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19120)

### CODE REVIEW

File: ports/glutin/window.rs [hunk 1]
```diff
@@ -470,81 +446,158 @@ impl Window {
...
+            Event::WindowEvent {
+                window_id: _,
+                event: glutin::WindowEvent::MouseInput {
+                    device_id: _,
+                    state: state,
+                    button: button,
+                },
+            } => {
+                // TODO: position is necessary according to @paulrouget, but pos is not given in
+                // the event's struct.
```
Comment (paulrouget) [inline ports/glutin/window.rs]: Yes. The PR that brings these coordinates to the wheel and click events didn't make it to upstream iirc.

File: ports/glutin/window.rs [hunk 2]
```diff
@@ -1040,74 +1064,64 @@ impl WindowMethods for Window {
...
-            WindowKind::Headless(..) => {}
+            },
+            WindowKind::Headless(..) => {},
         }
     }
 
     fn create_event_loop_waker(&self) -> Box<EventLoopWaker> {
         struct GlutinEventLoopWaker {
-            window_proxy: Option<glutin::WindowProxy>,
+            events_loop: Option<Arc<Mutex<glutin::EventsLoop>>>,
```
Comment (paulrouget) [inline ports/glutin/window.rs]: I don't think you need to use EventsLoop. Things will be probably easier with:

```
    proxy: Arc<glutin::EventsLoopProxy>,
```

File: ports/glutin/window.rs [hunk 3]
```diff
@@ -1040,74 +1064,64 @@ impl WindowMethods for Window {
...
         }
         impl EventLoopWaker for GlutinEventLoopWaker {
             fn wake(&self) {
                 // kick the OS event loop awake.
-                if let Some(ref window_proxy) = self.window_proxy {
-                    window_proxy.wakeup_event_loop()
+                if let Some(ref events_loop) = self.events_loop {
+                    events_loop.lock().unwrap().create_proxy().wakeup();
```
Comment (paulrouget) [inline ports/glutin/window.rs]: This will be easier if you use EventsLoopProxy. No need to create a proxy everytime.

**Comment by paulrouget** [comment]
An important change in upstream glutin is the way keys are handled.

If you look at servo glutin, there are 2 code paths, one for Mac+Linux, and one for Windows. Now, with upstream glutin, only the Windows code path is needed. For example, get rid of the `pending_key_event_char` logic.

**Comment by paulrouget** [comment]
For `GlutinEventLoopWaker`, I would recommend to copy what I did here:

https://github.com/paulrouget/servoshell/blob/13dc6fed301e9638b8fc60174f09fadd9f8fec7f/src/platform/glutin/app.rs#L21

**Comment by paulrouget** [comment]
About coordinates, please look at:

https://github.com/servo/glutin/pull/99
https://github.com/servo/glutin/pull/109

File: ports/glutin/window.rs [hunk 4]
```diff
@@ -410,52 +344,11 @@ impl Window {
...
-                ElementState::Released => KeyState::Released,
-            };
-            let modifiers = Window::glutin_mods_to_script_mods(self.key_modifiers.get());
-            self.event_queue.borrow_mut().push(WindowEvent::KeyEvent(ch, key, state, modifiers));
-        }
-    }
-
-    #[cfg(target_os = "windows")]
-    fn handle_keyboard_input(&self, element_state: ElementState, _scan_code: u8, virtual_key_code: VirtualKeyCode) {
+    fn set_key(&self, element_state: ElementState, virtual_key_code: VirtualKeyCode) {
```
Comment (stuartnelson3) [inline ports/glutin/window.rs]: I had no idea what to name this; any suggestion is welcome.

**Comment by stuartnelson3** [comment]
@paulrouget i looked at the two links you provided for getting mouse position, but I feel a bit stuck -- I'm not seeing a way in the glutin library to get anything other than the bounds of the window (inner or outer), and I couldn't see anything in the servo::gl library that would help, either.

any ideas?

**Comment by paulrouget** [comment]
The glutin library does not provide the mouse coordinates on click and wheel events.

So we need to fix the glutin library to do so. Now, glutin is split in 2: glutin for the opengl stuff, and winit for the windowing and events. So what you actually need to fix is winit.

The 2 links show how we did it for the servoglutin.

So someone need to adapt these 2 PRs to the winit library.

**Comment by jdm** [comment]
@stuartnelson3 What's still missing from the list in #17311?

**Comment by stuartnelson3** [comment]
@jdm I need to get winit bumped (https://github.com/tomaka/winit/pull/355), then update glutin, then include that here so that we can get the mouse coordinates whenever a user focuses the window. Currently, a user can lose focus on the servo window, move the mouse, then regain focus on the window without servo receiving any updated mouse position message. The change to winit fixes this.

**Comment by stuartnelson3** [comment]
- [ ] Get the icon loading patches either upstreamed or rebased on a new glutin fork.
- [ ] Get the platform_window() code working [or drop CEF for now?].

@jdm successfully got glutin bumped. The last steps are the above two, I believe. I'm not sure what how folks feel about completing the above two

**Comment by stuartnelson3** [comment]
Hm, the build is failing on `./mach test-tidy --no-progress --all`, not sure what to do about the duplicate packages it's complaining about.

**Comment by Eijebong** [comment]
@stuartnelson3 This patch needs to be upstreamed I think https://github.com/servo/glutin/commit/08518877e51a3e048c4de1c3396cd054b808ac91. I would do it myself but don't have any osx to test it. Let's ping @nox about it :)

**Comment by paulrouget** [comment]
@stuartnelson3 can you describe what's left to be done here? Do you need an early review?

**Comment by paulrouget** [comment]
Also, I'd be happy to take over if you don't have time to continue.

**Comment by stuartnelson3** [comment]
@paulrouget yeah feel free to take over. the patch in this comment needs to be upstreamed: https://github.com/servo/servo/pull/19120#issuecomment-349578613

Good luck!

**Comment by jdm** [comment]
@paulrouget I suspect we're still missing fixes for the work items listed in the original comment in https://github.com/servo/servo/pull/17311.

**Comment by paulrouget** [comment]
Taking over this PR. See #19895

@stuartnelson3 thanks a lot for your initial PR. That made my life a lot easier :)

**Pull request closed without merging**